lib/git_worktree.rb

```diff
@@ -53,18 +59,40 @@ def tags
     @repo.tags.each.collect(&:name)
   end
 
+  private def find_tag(name)
+    @repo.tags.each.detect { |t| t.name.casecmp(name) == 0 }
+  end
+
   def tag=(name)
-    tag = @repo.tags.each.detect { |t| t.name.casecmp(name) == 0 }
+    tag = find_tag(name)
     raise GitWorktreeException::TagMissing, name unless tag
     @commit_sha = tag.target.oid
   end
 
   def tag_info(name)
-    tag = @repo.tags.each.detect { |t| t.name.casecmp(name) == 0 }
+    tag = find_tag(name)
     raise GitWorktreeException::TagMissing, name unless tag
     {:time => tag.target.time, :message => tag.target.message, :commit_sha => tag.target.oid}
   end
 
+  private def find_ref(ref)
+    @repo.lookup(ref)
+  rescue Rugged::InvalidError, Rugged::OdbError
+    nil
+  end
+
+  def ref=(ref)
```

Comment:
Should `self.tag` and `self.branch` be mutually exclusive?  By that, I mean if you assign one, you unassign the other?

Reply:
Yes, those are the sematics of GitWorktree in that its purpose it to track something "current".  `ref=` is a more flexible method that allows branch or tag or any SHA.

Reply:
The point I was making is should this method do:

```diff
   def ref=(ref)
     if find_branch(ref)
       self.branch = ref
+      self.tag    = nil
     elsif find_tag(ref)
       self.tag    = ref
+      self.branch = nil
     elsif find_ref(ref)
       @commit_sha = @repo.lookup(ref).oid
     else
       raise GitWorktreeException::RefMissing, ref
     end
   end
```

As well?

Reply:
Ah no, that's not necessary...and of branch/tag/ref= set the commit_sha, and that is the expected way for it to work.  In fact, setting tag to nil after setting the branch will likely "erase" the branch that was just set.